nni/retiarii/experiment/config/experiment_config.py

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass
+from typing import Any
+
+from nni.experiment.config import utils, ExperimentConfig
+
+from .engine_config import ExecutionEngineConfig, PyEngineConfig
+
+__all__ = ['RetiariiExeConfig']
+
+def execution_engine_config_factory(engine_name):
+    # FIXME: may move this function to experiment utils in future
+    cls = _get_ee_config_class(engine_name)
+    if cls is None:
+        raise ValueError(f'Invalid execution engine name: {engine_name}')
+    return cls()
+
+def _get_ee_config_class(engine_name):
+    for cls in ExecutionEngineConfig.__subclasses__():
+        if cls.name == engine_name:
+            return cls
+    return None
+
+@dataclass(init=False)
+class RetiariiExeConfig(ExperimentConfig):
+    # FIXME: refactor this class to inherit from a new common base class with HPO config
+    search_space: Any = ''
+    trial_code_directory: utils.PathLike = '.'
+    trial_command: str = '_reserved'
+    # new config field for NAS
+    execution_engine: ExecutionEngineConfig = PyEngineConfig()
+
+    def __init__(self, training_service_platform: str | None = None,
+                 execution_engine: str | ExecutionEngineConfig = PyEngineConfig(),
+                 **kwargs):
+        super().__init__(training_service_platform, **kwargs)
+
+        if isinstance(execution_engine, str):
+            self.execution_engine = execution_engine_config_factory(execution_engine)
+        else:
+            self.execution_engine = execution_engine
+
+        self.__dict__['trial_command'] = 'python3 -m nni.retiarii.trial_entry ' + self.execution_engine.name
+
+    def __setattr__(self, key, value):
```

Comment:
updated

Reply:
it is not easy to do the checking in `_validate_canonical`, because the function `_canonicalize` may be executed multiple times accordingly. thus, still a little hacky, please look at the latest version.